File: apps/web/src/api/directory.ts
import { apiClient } from './client'

export interface DirectoryCard {
  slug: string
  display_name: string
  profession: string
  profession_label: string
  title: string | null
  city: string
  city_slug: string
  tier: string
  verified: boolean
  contactable: boolean
  photo_url: string | null
  headline: string | null
  focus_areas: string[]
  formats: string[]
  offers_free_intro: boolean
}

export interface DirectoryDetail extends DirectoryCard {
  postal_code: string | null
  state: string | null
  website: string | null
  phone: string | null
  about: string | null
  approach: string | null
  fees: string | null
  languages: string[]
  booking_url: string | null
  updated_at: string | null
}

export interface FacetItem {
  slug: string
  label: string
  count: number
}

export interface DirectoryFacets {
  total: number
  professions: FacetItem[]
  cities: FacetItem[]
}

export interface DirectorySearchResponse {
  total: number
  items: DirectoryCard[]
}

export interface DirectorySearchParams {
  q?: string
  profession?: string
  city?: string
  format?: string
  free_intro?: boolean
  page?: number
}

export interface DirectoryContactPayload {
  from_email: string
  from_name?: string
  from_phone?: string
  message?: string
  preferred_format?: string
  company?: string // Honeypot
}

export const directoryApi = {
  search: (params: DirectorySearchParams) =>
    apiClient.get<DirectorySearchResponse>('/directory/search', { params }).then((r) => r.data),
  facets: () => apiClient.get<DirectoryFacets>('/directory/facets').then((r) => r.data),
  detail: (slug: string) =>
    apiClient.get<DirectoryDetail>(`/directory/listings/${slug}`).then((r) => r.data),
  contact: (slug: string, payload: DirectoryContactPayload) =>
    apiClient
      .post<{ message: string }>(`/directory/listings/${slug}/contact`, payload)
      .then((r) => r.data),
}

// ── Selfservice-Profil (eingeloggte Fachperson) ──────────────────────────────

export interface MissingItem {
  key: string
  label: string
  points: number
}

export interface DirectoryMe {
  slug: string
  display_name: string
  profession: string
  title: string | null
  city: string
  postal_code: string | null
  state: string | null
  website: string | null
  phone: string | null
  contact_email: string | null
  photo_url: string | null
  headline: string | null
  about: string | null
  approach: string | null
  fees: string | null
  focus_areas: string[]
  formats: string[]
  languages: string[]
  offers_free_intro: boolean
  booking_url: string | null
  tier: string
  published: boolean
  completeness: number
  stars: number
  missing: MissingItem[]
  publishable: boolean
  missing_required: string[]
  public_url: string
}

export interface DirectoryProfilePayload {
  display_name: string
  profession: string
  title?: string | null
  city?: string | null
  postal_code?: string | null
  state?: string | null
  website?: string | null
  phone?: string | null
  contact_email?: string | null
  headline?: string | null
  about?: string | null
  approach?: string | null
  fees?: string | null
  focus_areas: string[]
  formats: string[]
  languages: string[]
  offers_free_intro: boolean
  booking_url?: string | null
  published: boolean
}

export const directoryProfileApi = {
  me: () => apiClient.get<DirectoryMe>('/directory/me').then((r) => r.data),
  save: (payload: DirectoryProfilePayload) =>
    apiClient.put<DirectoryMe>('/directory/me', payload).then((r) => r.data),
  uploadPhoto: (file: File) => {
    const fd = new FormData()
    fd.append('file', file)
    return apiClient
      .post<{ photo_url: string }>('/directory/me/photo', fd, {
        headers: { 'Content-Type': 'multipart/form-data' },
        timeout: 60_000,
      })
      .then((r) => r.data)
  },
}

File: services/api/app/schemas/directory.py
"""Schemas fürs öffentliche Fachpersonen-Verzeichnis ("Fachperson finden")."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, EmailStr, Field


class DirectoryCard(BaseModel):
    """Kompakter Eintrag für Suchergebnis-/Regionallisten."""
    slug: str
    display_name: str
    profession: str
    profession_label: str
    title: str | None = None
    city: str
    city_slug: str
    tier: str
    verified: bool = False
    contactable: bool = False        # tier != researched
    photo_url: str | None = None
    headline: str | None = None
    focus_areas: list[str] = []
    formats: list[str] = []
    offers_free_intro: bool = False


class DirectoryDetail(DirectoryCard):
    """Vollständige Profilseite. contact_email wird bewusst NIE ausgeliefert
    (Weiterleitung nur serverseitig, Schutz vor Scraping)."""
    postal_code: str | None = None
    state: str | None = None
    website: str | None = None
    phone: str | None = None
    about: str | None = None
    approach: str | None = None
    fees: str | None = None
    languages: list[str] = []
    booking_url: str | None = None
    updated_at: datetime | None = None


class FacetItem(BaseModel):
    slug: str
    label: str
    count: int


class DirectoryFacets(BaseModel):
    total: int
    professions: list[FacetItem]
    cities: list[FacetItem]


class DirectorySearchResponse(BaseModel):
    total: int
    items: list[DirectoryCard]


class DirectoryContactCreate(BaseModel):
    from_email: EmailStr
    from_name: str | None = Field(default=None, max_length=120)
    from_phone: str | None = Field(default=None, max_length=40)
    message: str | None = Field(default=None, max_length=3000)
    preferred_format: str | None = Field(default=None, max_length=20)
    company: str | None = None       # Honeypot – von Bots ausgefüllt


class ContactAck(BaseModel):
    message: str


# ── Selfservice-Profil (eingeloggte Fachperson) ──────────────────────────────

class DirectoryProfileUpdate(BaseModel):
    display_name: str = Field(min_length=1, max_length=120)
    profession: str = ""
    title: str | None = None
    city: str | None = None
    postal_code: str | None = None
    state: str | None = None
    website: str | None = None
    phone: str | None = None
    contact_email: str | None = None
    headline: str | None = Field(default=None, max_length=160)
    about: str | None = None
    approach: str | None = None
    fees: str | None = None
    focus_areas: list[str] = []
    formats: list[str] = []
    languages: list[str] = []
    offers_free_intro: bool = False
    booking_url: str | None = None
    published: bool = False


class MissingItem(BaseModel):
    key: str
    label: str
    points: int


class DirectoryMe(BaseModel):
    """Eigene Sicht der Fachperson auf ihr Listing — inkl. contact_email + Fortschritt."""
    slug: str
    display_name: str
    profession: str
    title: str | None = None
    city: str = ""
    postal_code: str | None = None
    state: str | None = None
    website: str | None = None
    phone: str | None = None
    contact_email: str | None = None
    photo_url: str | None = None
    headline: str | None = None
    about: str | None = None
    approach: str | None = None
    fees: str | None = None
    focus_areas: list[str] = []
    formats: list[str] = []
    languages: list[str] = []
    offers_free_intro: bool = False
    booking_url: str | None = None
    tier: str
    published: bool
    completeness: int
    stars: int
    missing: list[MissingItem]
    publishable: bool
    missing_required: list[str]
    public_url: str


class PhotoResult(BaseModel):
    photo_url: str
